lucene/core/src/java/org/apache/lucene/util/graph/GraphTokenStreamFiniteStrings.java

```diff
@@ -0,0 +1,237 @@
+    this.builder = new Automaton.Builder();
+  }
+
+  private static class BytesRefArrayTokenStream extends TokenStream {
+    private final BytesTermAttribute termAtt = addAttribute(BytesTermAttribute.class);
+    private final PositionIncrementAttribute posIncAtt = addAttribute(PositionIncrementAttribute.class);
+
+    private final BytesRef[] terms;
+    private final int[] increments;
+    private int offset;
+
+    BytesRefArrayTokenStream(BytesRef[] terms, int[] increments) {
+      this.terms = terms;
+      this.increments = increments;
+      assert terms.length == increments.length;
+      offset = 0;
+    }
+
+    @Override
+    public boolean incrementToken() throws IOException {
+      if (offset < terms.length) {
+        clearAttributes();
+        termAtt.setBytesRef(terms[offset]);
+        posIncAtt.setPositionIncrement(increments[offset]);
+        offset = offset + 1;
+        return true;
+      }
+
+      return false;
+    }
+  }
+
+  /**
+   * Gets the list of finite string token streams from the given input graph token stream.
+   */
+  public List<TokenStream> getTokenStreams(final TokenStream in) throws IOException {
+    // build automation
+    build(in);
+
+    List<TokenStream> tokenStreams = new ArrayList<>();
+    final FiniteStringsIterator finiteStrings = new FiniteStringsIterator(det);
+    for (IntsRef string; (string = finiteStrings.next()) != null; ) {
+      final BytesRef[] tokens = new BytesRef[string.length];
+      final int[] increments = new int[string.length];
+      for (int idx = string.offset, len = string.offset + string.length; idx < len; idx++) {
+        int id = string.ints[idx];
+        int offset = idx - string.offset;
+        tokens[offset] = idToTerm.get(id);
+        if (idToInc.containsKey(id)) {
+          increments[offset] = idToInc.get(id);
+        } else {
+          increments[offset] = 1;
+        }
+      }
+
+      tokenStreams.add(new BytesRefArrayTokenStream(tokens, increments));
+    }
+
+    return tokenStreams;
+  }
+
+  private void build(final TokenStream in) throws IOException {
+    if (det != null) {
+      throw new IllegalStateException("Automation already built");
+    }
+
+    final TermToBytesRefAttribute termBytesAtt = in.addAttribute(TermToBytesRefAttribute.class);
+    final PositionIncrementAttribute posIncAtt = in.addAttribute(PositionIncrementAttribute.class);
+    final PositionLengthAttribute posLengthAtt = in.addAttribute(PositionLengthAttribute.class);
+    final OffsetAttribute offsetAtt = in.addAttribute(OffsetAttribute.class);
+
+    in.reset();
+
+    int pos = -1;
+    int lastIncr = 1;
+    int maxOffset = 0;
+    int maxPos = -1;
+    int state = -1;
+    while (in.incrementToken()) {
+      int posInc = posIncAtt.getPositionIncrement();
+
+      // always use inc 1 while building, but save original increment
+      int fakePosInc = posInc > 1 ? 1 : posInc;
+
+      assert pos > -1 || fakePosInc > 0;
```

Comment:
Can we upgrade this to a real `if`?  I.e. we need a well-formed `TokenStream` input ... it cannot have `posInc=0` on its first token.